### PR TITLE
Allow query to display while slice is loading

### PR DESCRIPTION
**Pull request by vera-liu**
Done:
 - separated get_query_str out from query() interface for sqlalchemy and druid
 - pass in queryEndpoint instead of query to DisplayQueryButton
 - add query=true option for explore_json to fetch query string without doing the query
 - get query_str from explore_json when query button is displayed

After:
user could still see the most updated query str from latest form_data even when slice hasn't done rendering:
screen shot 2017-02-02 at 3 08 58 pm

Todo:
test in druid

needs-review @mistercrunch

### CODE REVIEW

File: superset/viz.py [hunk 1]
```diff
@@ -106,10 +106,11 @@ def get_df(self, query_obj=None):
                 timestamp_format = dttm_col.python_date_format
 
         # The datasource here can be different backend but the interface is common
-        self.results = self.datasource.query(**query_obj)
+        query_str, engine, start_dttm = self.datasource.get_query_str(**query_obj)
```
Comment (mistercrunch) [inline superset/viz.py]: At first sight it seems like you wouldn't need start_dttm and engine to be returned by `get_query_str`. Functions returning tuples are generally a bad idea, it points in the direction of making a more targeted function or returning an object. In this case the former.

File: superset/viz.py [hunk 2]
```diff
@@ -106,10 +106,11 @@ def get_df(self, query_obj=None):
...
+        self.query = query_str
+        self.results = self.datasource.query(query_str, engine, start_dttm)
```
Comment (mistercrunch) [inline superset/viz.py]: The engine can probably be fetched within the scope of that function, and start_dttm can probably generated there also.

File: superset/viz.py [hunk 3]
```diff
@@ -312,6 +313,11 @@ def get_csv(self):
         include_index = not isinstance(df.index, pd.RangeIndex)
         return df.to_csv(index=include_index, encoding="utf-8")
 
+    def get_query(self):
```
Comment (mistercrunch) [inline superset/viz.py]: This method shouldn't be needed if the other ones are better designed.

**Review by mistercrunch — CHANGES_REQUESTED** [review]
I had forgotten that Druid has some extra complexity with 2-phases queries, meaning that figuring out the full query requires running a query. For that specific case we may need to only return phase1 if the query is currently executing.

**Comment by vera-liu** [comment]
Removed tuple return value and added phase one query for druid, tested in staging for sqla/druid. PTAL @mistercrunch

**Comment by mistercrunch** [comment]
Ooops merge conflict. LGTM otherwise

**Comment by vera-liu** [comment]
Resolved conflicts 👍

**Pull request merged**